### PR TITLE
Add IXWebSocket 9.1.9

### PR DESCRIPTION
**Pull request by LunarWatcher**
Specify library name and version:  **ixwebsocket/9.1.9**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

Only tested locally against a Linux build. This version seems to cooperate with the existing code. Windows might fight back and cause the build to fail. (This version is greater than the version threshold for enabling OpenSSL on Windows; it might fail. Unexplored territory, and I don't have Windows available for verification atm)

Nothing seems to have broken significantly anyway.

### CODE REVIEW

**Comment by LunarWatcher** [comment]
@uilianries That will fail the build though. The initial change only changed those two.

Should 9.1.9 and up be split into a separate folder instead? Guessing the problem is the change of an already published version?

File: recipes/ixwebsocket/all/conanfile.py [hunk 1]
```diff
@@ -28,6 +28,8 @@ class IXWebSocketConan(ConanFile):
         "use_openssl": [False, True],
         "fPIC": [True, False]
     }
+
+
```
Comment (uilianries) [inline recipes/ixwebsocket/all/conanfile.py:32]:
```suggestion
```

File: recipes/ixwebsocket/all/conanfile.py [hunk 2]
```diff
@@ -43,8 +45,9 @@ def configure(self):
         if self.options.use_mbed_tls and not self.options.use_tls or self.options.get_safe("use_openssl") and not self.options.use_tls:
             raise ConanInvalidConfiguration("TLS must be enabled to use mbedtls")
         elif self.options.use_mbed_tls and self.options.get_safe("use_openssl"):
-            raise ConanInvalidConfiguration("Cannot use both OpenSSL and MbedTLS")
+            raise ConanInvalidConfiguration("Cannot use both OpenSSL and MbedTLS")
```
Comment (uilianries) [inline recipes/ixwebsocket/all/conanfile.py:48]:
```suggestion
            raise ConanInvalidConfiguration("Cannot use both OpenSSL and MbedTLS")
```

trailing space

File: recipes/ixwebsocket/all/conanfile.py (hunk 1 above)
Comment (uilianries) [inline recipes/ixwebsocket/all/conanfile.py:32]:
```suggestion
```

File: recipes/ixwebsocket/all/conanfile.py [hunk 3]
```diff
@@ -43,8 +45,9 @@ def configure(self):
         if self.options.use_mbed_tls and not self.options.use_tls or self.options.get_safe("use_openssl") and not self.options.use_tls:
             raise ConanInvalidConfiguration("TLS must be enabled to use mbedtls")
         elif self.options.use_mbed_tls and self.options.get_safe("use_openssl"):
-            raise ConanInvalidConfiguration("Cannot use both OpenSSL and MbedTLS")
+            raise ConanInvalidConfiguration("Cannot use both OpenSSL and MbedTLS") 
 
+
```
Comment (uilianries) [inline recipes/ixwebsocket/all/conanfile.py:50]:
```suggestion
```

File: recipes/ixwebsocket/all/conanfile.py [hunk 4]
```diff
@@ -56,16 +59,20 @@ def config_options(self):
...
             self.requires.add("openssl/1.1.1c")
 
         self.requires.add("zlib/1.2.11")
 
+        # Windows requires a bit of manual jiggling for TLS fallbacks.  
+        # Apple will fall back on its own SSL system of not using mbed or OpenSSL.
         if self.settings.os == "Windows" and self.options.use_tls and not self._can_use_openssl() or self.options.use_mbed_tls:
             self.requires.add("mbedtls/2.16.3-apache")
 
+
```
Comment (uilianries) [inline recipes/ixwebsocket/all/conanfile.py:75]:
```suggestion
```

File: recipes/ixwebsocket/all/conanfile.py [hunk 5]
```diff
@@ -56,16 +59,20 @@ def config_options(self):
             del self.options.use_openssl
 
     def requirements(self):
+        if ((self._can_use_openssl() and not self.options.use_mbed_tls and self.options.use_tls
+                and (self.options.get_safe("use_openssl") or self.settings.os == "Linux" 
+                     or self.settings.os == "Windows" and not self.options.get_safe("openssl")))): # OpenSSL jiggling with Windows necessary on newer versions.
```
Comment (uilianries) [inline recipes/ixwebsocket/all/conanfile.py:64]:
```suggestion
        if self._can_use_openssl() and not self.options.use_mbed_tls and self.options.use_tls
                and (self.options.get_safe("use_openssl") or self.settings.os == "Linux" 
                     or self.settings.os == "Windows" and not self.options.get_safe("openssl"): # OpenSSL jiggling with Windows necessary on newer versions. 
```

File: recipes/ixwebsocket/all/conanfile.py [hunk 6]
```diff
@@ -56,16 +59,20 @@ def config_options(self):
             del self.options.use_openssl
 
     def requirements(self):
+        if ((self._can_use_openssl() and not self.options.use_mbed_tls and self.options.use_tls
+                and (self.options.get_safe("use_openssl") or self.settings.os == "Linux" 
+                     or self.settings.os == "Windows" and not self.options.get_safe("openssl")))): # OpenSSL jiggling with Windows necessary on newer versions. 
 
-        if(self._can_use_openssl() and not self.options.use_mbed_tls and self.options.use_tls
-           and (self.options.get_safe("use_openssl") or self.settings.os == "Linux")):
             self.requires.add("openssl/1.1.1c")
```
Comment (uilianries) [inline recipes/ixwebsocket/all/conanfile.py:66]:
```suggestion
            self.requires.add("openssl/1.1.1e")
```

File: recipes/ixwebsocket/all/conanfile.py [hunk 7]
```diff
@@ -56,13 +57,16 @@ def config_options(self):
             del self.options.use_openssl
 
     def requirements(self):
+        if (self._can_use_openssl() and not self.options.use_mbed_tls and self.options.use_tls
+                and (self.options.get_safe("use_openssl") or self.settings.os == "Linux" 
+                     or self.settings.os == "Windows" and not self.options.get_safe("openssl"))): # OpenSSL jiggling with Windows necessary on newer versions.
```
Comment (jgsogo) [inline recipes/ixwebsocket/all/conanfile.py:62]: There is no `openssl` option.

Also, can we add a parenthesis here, not sure what is the intention of the last `and`...

**Comment by LunarWatcher** [comment]
@jgsogo To make a long story short, different versions have caused the system to be a bit of a mess. [This is the current system](https://github.com/machinezone/IXWebSocket/blob/master/CMakeLists.txt#L129-L139); mbed takes priority over openssl and apple's SSL system. Apple's SSL system takes priority over OpenSSL, and OpenSSL is the fallback. The oldest version registered to conan (7.9.2) actually uses [another system](https://github.com/machinezone/IXWebSocket/blob/v9.1.9/CMakeLists.txt#L130-L140) where Windows doesn't use OpenSSL at all. Windows in 7.9.2 requires MbedTLS. 

A picture says a thousand words, so here's a flowchart:

image



openssl instead of use_openssl is a typo (nice catch). There's also a few logic errors actually (the entire dependency graph makes this extremely confusing to deal with), which I'm working on right now.

Somewhat annoyingly, a newer version actually enforces mbed if neither openssl or mbed is defined. For the next version, I'm splitting it into a new folder, because I cannot maintain readability at this rate. Additionally, this change will be the first to break with the current version of the conanfile. It might even be better at this point to abort and revert, and fork the current changes into a separate file to avoid some of the spaghetti.

**Comment by jgsogo** [comment]
Wonderful diagram, thanks a lot!


> Somewhat annoyingly, a newer version actually enforces mbed if neither openssl or mbed is defined. For the next version, I'm splitting it into a new folder, because I cannot maintain readability at this rate. Additionally, this change will be the first to break with the current version of the conanfile. It might even be better at this point to abort and revert, and fork the current changes into a separate folder to avoid some of the spaghetti.

I'd vote for a forked recipe in a separate folder if it helps with readability and to maintain future versions of this recipe. The old one can be there and get some fixes if people report a failure.

**Comment by LunarWatcher** [comment]
Alright, I'll add a revert commit and implement a new version separately. I'll push them at the same time to avoid wasted CI time. Currently working on your suggestion to implement a list of TLS libraries as well, but I'm probably only gonna do that in the new recipe. this approach has a few consequences with flag calculation (although, tbf, so does the original. We'll see). That's causing some challenges to, though, mainly with flags, combined with several mistakes from the initial changes. 

I'm also gonna declutter the future implementation stuff from the earliest recipe just to make it cleaner. Maintaining this recipe with a single file ended up being more complicated than I expected ^^"

I'll see where this ends up, I guess

**Comment by LunarWatcher** [comment]
Implementing a list wasn't as bad as I thought. Was thinking a lot more complicated before I actually implemented it. I had to cross-reference with some other recipes to figure out how they dealt with implementations, and the ones I checked used configure to get rid of invalid setting combinations instead of deleting some and requiring spaghetti checks with variables that may or may not exist later to determine fallbacks. It makes it possible to use a single folder for now (in theory:tm: - it works on Linux. Windows is still the wildcard, and _maybe_ mac?).

The reason I'm not doing it in different folders is because this approach actually enforces one specific framework. This way, instead of dealing with dependencies in case the flags may allow for a fallback use, it's determined by the packaging script instead. One difference is on Windows with 7.9.2, where mbedtls is used independently of settings, and overruling the Windows fallback. But for this version, _can_use_openssl returns false, and anything aside mbedtls will throw an invalid configuration error. 

It's also easy to split the recipe later if necessary (read: if 7.9.2 fails and the patch requires changes that are specific to that version, it's easier to split the recipe instead of mixing in spaghetti code). The only difference is that can_use_openssl needs to be replaced with an OS check for Windows on 7.9.2 and earlier, and the entire branch can be removed for later versions.

TL;DR: the script determines fallbacks now instead of dealing with spaghetti checks to let IXWebSocket's CMake file determine them with possible dependency failures.

File: recipes/ixwebsocket/all/conanfile.py [hunk 8]
```diff
@@ -41,33 +39,33 @@ def _can_use_openssl(self):
...
-        if self.options.use_mbed_tls and not self.options.use_tls or self.options.get_safe("use_openssl") and not self.options.use_tls:
-            raise ConanInvalidConfiguration("TLS must be enabled to use mbedtls")
-        elif self.options.use_mbed_tls and self.options.get_safe("use_openssl"):
-            raise ConanInvalidConfiguration("Cannot use both OpenSSL and MbedTLS")
+        if self.options.tls == "applessl" and not tools.is_apple_os(self.settings.os):
+            raise ConanInvalidConfiguration("Can only use Apple SSL on Apple.")
+        elif self.options.tls and not self.options.use_tls:
+            raise ConanInvalidConfiguration("You cannot use an SSL system when use_tls is false")
+        elif not self.options.tls and self.options.use_tls:
+            raise ConanInvalidConfiguration("You need to select an SSL system when use_tls is true")
```
Comment (jgsogo) [inline recipes/ixwebsocket/all/conanfile.py:47]: Given the two conditions before this line, I wonder if the `use_tls` option is needed. The value `None` for `tls` can work the same as "use_tls=False", wdyt?

Comment (LunarWatcher) [reply recipes/ixwebsocket/all/conanfile.py:47]: Yeah, that is reasonable. I'll get rid of  `use_tls`

File: recipes/ixwebsocket/all/conanfile.py [hunk 9]
```diff
@@ -40,30 +38,29 @@ def _can_use_openssl(self):
         return True
 
     def configure(self):
-        if self.options.use_mbed_tls and not self.options.use_tls or self.options.get_safe("use_openssl") and not self.options.use_tls:
-            raise ConanInvalidConfiguration("TLS must be enabled to use mbedtls")
-        elif self.options.use_mbed_tls and self.options.get_safe("use_openssl"):
-            raise ConanInvalidConfiguration("Cannot use both OpenSSL and MbedTLS")
+        if self.options.tls == "applessl" and not tools.is_apple_os(self.settings.os):
```
Comment (uilianries) [inline recipes/ixwebsocket/all/conanfile.py:41]: it looks much simpler now :+1:
